ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto

```diff
@@ -12,30 +12,60 @@ option java_outer_classname = "TransactionFilterOuterClass";
 option java_package = "com.daml.ledger.api.v1";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1";
 
-// Used for filtering Transaction and Active Contract Set streams.
-// Determines which on-ledger events will be served to the client.
+// A filter both for filtering create and archive events as well as for
+// filtering transaction trees.
 message TransactionFilter {
 
-  // Keys of the map determine which parties' on-ledger transactions are being queried.
-  // Values of the map determine which events are disclosed in the stream per party.
-  // At the minimum, a party needs to set an empty Filters message to receive any events.
   // Each key must be a valid PartyIdString (as described in ``value.proto``).
+  // The interpretation of the filter depends on the stream being filtered:
+  // (1) For **transaction tree streams** only party filters with wildcards are allowed, and all subtrees
+  //     whose root has one of the listed parties as an informee are returned.
+  // (2) For **transaction and active-contract-set streams** create and archive events are returned for all contracts whose
+  //    stakeholders include at least one of the listed parties and match the
+  //    per-party filter.
   // Required
   map<string, Filters> filters_by_party = 1;
 }
 
+// The union of a set of contract filters, or a wildcard.
 message Filters {
 
-  // If not set, no filters will be applied.
+  // If set, then contracts matching any of the ``InclusiveFilters`` match
+  // this filter.
+  // If not set, any contract matches this filter.
   // Optional
   InclusiveFilters inclusive = 1;
 }
 
-// If no internal fields are set, no filters will be applied.
+// A filter that matches all contracts that are either an instance of one of
+// the ``template_ids`` or that match one of the ``interface_filters``.
 message InclusiveFilters {
 
-  // A collection of templates.
+  // A collection of templates for which the payload will be included in the
+  // ``create_arguments`` of a matching ``CreatedEvent``.
   // SHOULD NOT contain duplicates.
-  // Required
+  // All ``template_ids`` needs to be valid: corresponding template should be defined in one of
+  // the available packages at the time of the query.
+  // Optional
   repeated Identifier template_ids = 1;
+
+  // Include an ``InterfaceView`` for every ``InterfaceFilter`` matching a contract.
+  // The ``InterfaceFilter``s MUST use unique ``interface_id``s.
```

Comment:
the MUST uniqueness's scope is per InclusiveFilters, or per filters_by_party?
It seems to me if have same interface_id's for two parties, and witness fires for both, then we are back with the merging problem.
Also if we have even just unique interface_id InterfaceFilter definitions, but have a template ID for which multiple interface_id applies, we are back again with merging InterfaceFilter.
So it seems to me merging Filters for rendering view-s is not easy to avoid.
Rules for that seems straightforward for me: or those predicates (either of them is true then we include). WDYT?
If merging is unavoidable, I suggest documenting merging rules in the API as well.
Does this change our view with the `MUST`?
@stefanobaghino-da @meiersi-da @moritzkiefer-da

Reply:
On a second thought: since we are defining data-rendering properties in filters, would it make more sense to invert the relation and switch from Map[party, Set[filter-with-rendering]] to a tuple of (Map[filter-with-rendering, Set[party]], Set[party]), the second for wildcard parties.
Or just separating filtration from rendering entirely (in implementation, it is separated anyway: we get some event, take the template_id, and based on that we lookup the rendering): Map[party, Set[filter-without-rendering]] and Map[interface-id, rendering]

Reply:
Great question, I had forgotten about multiple parties. 

I think you’re right that we cannot really avoid merging of rendering properties so I’d try to not prevent duplicates and instead document merging.

There is one case that we want to pay attention to when merging:
If I have two filters for the same interface id for the same party then I just include all properties requested in those filters.
However if those two filters are for the same interface id but for different parties, things are slightly more complex: If both parties are stakeholders we do the same as before. But if only one party is a stakeholder we should probably only include the render properties from that filter rather than from both.

> On a second thought: since we are defining data-rendering properties in filters, would it make more sense to invert the relation and switch from Map[party, Set[filter-with-rendering]] to a tuple of (Map[filter-with-rendering, Set[party]], Set[party]), the second for wildcard parties.
> Or just separating filtration from rendering entirely (in implementation, it is separated anyway: we get some event, take the template_id, and based on that we lookup the rendering): Map[party, Set[filter-without-rendering]] and Map[interface-id, rendering]

In the first sketch of the API I built, I tried to keep a clearer distinction between filters & rendering properties. However, the API got pretty complex and @meiersi-da convinced me that this simpler representation is preferable: It’s closer to the template API, it leaves out some edge cases like rendering to an interface that is different from the filter & while there are some edge cases like the ones you pointed out, overall I suspect it might still be easier for users to understand.

Reply:
In the first sketch of the API I built, I tried to keep a clearer distinction between filters & rendering properties. However, the API got pretty complex and @meiersi-da convinced me that this simpler representation is preferable: It’s closer to the template API, it leaves out some edge cases like rendering to an interface that is different from the filter & while there are some edge cases like the ones you pointed out, overall I suspect it might still be easier for users to understand.

Thanks, it makes sense, I missed these iterations. We can make it work with this API, all good +1

    However if those two filters are for the same interface id but for different parties, things are slightly more complex: If both parties are stakeholders we do the same as before. But if only one party is a stakeholder we should probably only include the render properties from that filter rather than from both.

Does that mean that we want to implement merging not only per template, but also per template and witness?
Like:

We populate an interface view I for a Create Node with template Id T with witnesses P1 P2 P3,
if and only if, an InterfaceFilter with PFILTER party and IFILTER interface exists for:

    IFILTER == I
    and PFILTER is either P1, P2, P3
    and include_interface_view is true
    and T implements I

We populate create_arguments for a Create Node with template Id T with witnesses P1 P2 P3 if and only if:

    A wildcard filter exists for party either P1, P2, P3,
    Or a template filter exists with template T and with party either P1, P2, P3

Reply:
Yes that matches what I was trying to suggest. Maybe one comment on wildcard filters: The mental model I use is that wildcard filters are syntax sugar for passing in all template ids. That’s not completely accurate because they pick up templates uploaded later but as a guideline I think it gives us the semantics we usually want.

Do you see any issues implementing those semantics? If you have suggestions for simpler semantics, I’m also not opposed to it. This seems like the API I’d expect from a user pov but if we can tweak it a bit to significantly simplify the implementation that’s definitely worth considering.

Reply:
I don't see big issues with this, let's try to evolve the API in this direction.

Reply:
this is done in the latest commit